File: cloudflare/src/targets.ts
import { compileBrandRegex, matchBrand } from "./brand-match.ts";
import { isGroundingRedirectUrl } from "./citations.ts";
import type {
  Citation,
  MonitorTarget,
  ObservationState,
  PropertyResult,
  ProviderRun,
  RunRequest,
  Source,
} from "./types.ts";

export const MAX_MONITOR_PROPERTIES = 5;

export const CATEGORY_LABELS: Record<PropertyResult["category"], string> = {
  owned: "Owned",
  of_interest: "Of interest",
  competition: "Competition",
};

const CATEGORY_ALIASES: Record<string, PropertyResult["category"]> = {
  owned: "owned",
  of_interest: "of_interest",
  interest: "of_interest",
  competition: "competition",
  competitor: "competition",
  competitors: "competition",
};

export function displayLabel(target: MonitorTarget) {
  return String(target.label || "").trim() || target.value;
}

export function parseTargetCategory(value: unknown): PropertyResult["category"] {
  const normalized = String(value || "owned")
    .trim()
    .toLowerCase()
    .replace(/-/g, "_")
    .replace(/\s+/g, "_");
  return CATEGORY_ALIASES[normalized] || "owned";
}

function validateTargetValue(value: string) {
  try {
    new URL(value.includes("://") ? value : `https://${value}`);
  } catch {
    throw new Error(`Property "${value}" must be a valid domain or HTTP(S) URL.`);
  }
}

export function normalizeTargets(raw: MonitorTarget[]): MonitorTarget[] {
  const normalized: MonitorTarget[] = [];
  const seen = new Set<string>();
  for (const target of raw) {
    const value = String(target.value || "").trim().slice(0, 2_000);
    if (!value) continue;
    const key = value.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    validateTargetValue(value);
    const brandRegex = String(target.brandRegex || "").trim();
    compileBrandRegex(brandRegex);
    normalized.push({
      value,
      matchMode: target.matchMode || "root_domain",
      label: String(target.label || "").trim().slice(0, 120),
      category: parseTargetCategory(target.category),
      brandRegex: brandRegex || undefined,
    });
  }
  return normalized;
}

export function parseRunTargets(body: Partial<RunRequest>): MonitorTarget[] {
  if (Array.isArray(body.targets) && body.targets.length) {
    return normalizeTargets(body.targets);
  }
  const legacyValue = String(body.target || "").trim();
  if (!legacyValue) {
    throw new Error("Add at least one property to monitor.");
  }
  validateTargetValue(legacyValue);
  const brandRegex = String(body.brandRegex || "").trim();
  compileBrandRegex(brandRegex);
  const matchMode = ["root_domain", "exact_hostname", "url_prefix"].includes(String(body.matchMode))
    ? (body.matchMode as MonitorTarget["matchMode"])
    : "root_domain";
  return normalizeTargets([
    {
      value: legacyValue,
      matchMode,
      category: "owned",
      brandRegex: brandRegex || undefined,
    },
  ]);
}

export function validateRunTargets(targets: MonitorTarget[]) {
  if (!targets.length) throw new Error("Add at least one property to monitor.");
  if (targets.length > MAX_MONITOR_PROPERTIES) {
    throw new Error(`At most ${MAX_MONITOR_PROPERTIES} properties are supported.`);
  }
  return targets;
}

const MULTI_PART_SUFFIXES = new Set([
  "co.uk",
  "org.uk",
  "ac.uk",
  "com.au",
  "co.jp",
  "com.br",
  "co.nz",
  "com.mx",
]);

function rootDomain(hostname: string) {
  const normalized = hostname.toLowerCase().replace(/\.$/, "");
  const parts = normalized.split(".").filter(Boolean);
  if (parts.length <= 2) return normalized;
  const lastTwo = parts.slice(-2).join(".");
  if (MULTI_PART_SUFFIXES.has(lastTwo) && parts.length >= 3) {
    return parts.slice(-3).join(".");
  }
  return lastTwo;
}

export function targetMatchesTarget(target: MonitorTarget, candidate: string) {
  try {
    const candidateUrl = new URL(candidate.includes("://") ? candidate : `https://${candidate}`);
    const targetUrl = new URL(
      target.value.includes("://") ? target.value : `https://${target.value}`,
    );
    if (target.matchMode === "exact_hostname") {
      return candidateUrl.hostname.toLowerCase() === targetUrl.hostname.toLowerCase();
    }
    if (target.matchMode === "url_prefix") {
      return (
        candidateUrl.origin === targetUrl.origin &&
        (candidateUrl.pathname === targetUrl.pathname ||
          candidateUrl.pathname.startsWith(`${targetUrl.pathname.replace(/\/$/, "")}/`))
      );
    }
    return rootDomain(candidateUrl.hostname) === rootDomain(targetUrl.hostname);
  } catch {
    return false;
  }
}

export function matchingTargets(request: RunRequest, candidate: string) {
  return request.targets.filter((target) => targetMatchesTarget(target, candidate)).map((target) => target.value);
}

export function targetMatchFields(request: RunRequest, candidate: string) {
  const targetMatches = matchingTargets(request, candidate);
  return {
    targetMatch: targetMatches.length > 0,
    targetMatches,
  };
}

function aggregateState(
  results: PropertyResult[],
  field: keyof Pick<PropertyResult, "retrieved" | "cited" | "brandMentioned">,
) {
  if (!results.length) return "UNKNOWN" as ObservationState;
  const states = results.map((item) => item[field]);
  if (states.some((state) => state === "YES")) return "YES";
  if (states.some((state) => state === "UNKNOWN")) return "UNKNOWN";
  if (states.every((state) => state === "N/A")) return "N/A";
  return "NO";
}

export function computePropertyResults(
  run: ProviderRun,
  request: RunRequest,
  options: { retrievalComplete?: boolean; citationComplete?: boolean } = {},
): PropertyResult[] {
  const retrievalComplete = !!options.retrievalComplete;
  const citationComplete = options.citationComplete !== false;
  return request.targets.map((target) => {
    const hasFailedRedirects = run.citations.some(
      (citation) => isGroundingRedirectUrl(citation.url) && citation.redirectResolution === "failed",
    );
    const shouldResolve =
      request.resolveCitationRedirects ??
      request.targets.some((item) => item.matchMode === "url_prefix");
    const retrieved = run.sources.some((source) => source.targetMatches.includes(target.value))
      ? "YES"
      : retrievalComplete
        ? "NO"
        : "UNKNOWN";
    const cited = run.citations.some((citation) => citation.targetMatches.includes(target.value))
      ? "YES"
      : target.matchMode === "url_prefix" && shouldResolve && hasFailedRedirects
        ? "UNKNOWN"
        : citationComplete
          ? "NO"
          : "UNKNOWN";
    const brand = matchBrand(run.responseText, target.brandRegex);
    return {
      value: target.value,
      label: displayLabel(target),
      category: target.category,
      retrieved,
      cited,
      brandMentioned: brand.state,
      brandMatches: brand.matches,
    };
  });
}

export function applyPropertyResults(
  run: ProviderRun,
  request: RunRequest,
  options: { retrievalComplete?: boolean; citationComplete?: boolean } = {},
) {
  const propertyResults = computePropertyResults(run, request, options);
  run.propertyResults = propertyResults;
  run.targetRetrieved = aggregateState(propertyResults, "retrieved");
  run.targetCited = aggregateState(propertyResults, "cited");
  run.metadata.propertyResults = propertyResults;
  return run;
}

export function mergeTargetMatchFields(
  existing: Pick<Source | Citation, "targetMatch" | "targetMatches">,
  incoming: Pick<Source | Citation, "targetMatch" | "targetMatches">,
) {
  existing.targetMatches = [...new Set([...existing.targetMatches, ...incoming.targetMatches])];
  existing.targetMatch = existing.targetMatches.length > 0;
}

File: core/targets.py
from __future__ import annotations

from .brand_match import compile_brand_regex, match_brand
from .citation_redirects import should_resolve_citation_redirects
from .enums import MatchMode, ObservationState, TargetCategory
from .matching import is_grounding_redirect_url
from .models import GroundingRequest, GroundingRun, PropertyResult, Target

MAX_MONITOR_PROPERTIES = 5
CATEGORY_LABELS = {
    TargetCategory.OWNED: "Owned",
    TargetCategory.OF_INTEREST: "Of interest",
    TargetCategory.COMPETITION: "Competition",
}


def display_label(target: Target) -> str:
    return (target.label or "").strip() or target.value


def parse_target_category(value: str | TargetCategory | None) -> TargetCategory:
    if isinstance(value, TargetCategory):
        return value
    normalized = str(value or TargetCategory.OWNED.value).strip().lower().replace("-", "_").replace(" ", "_")
    aliases = {
        "owned": TargetCategory.OWNED,
        "of_interest": TargetCategory.OF_INTEREST,
        "interest": TargetCategory.OF_INTEREST,
        "competition": TargetCategory.COMPETITION,
        "competitor": TargetCategory.COMPETITION,
        "competitors": TargetCategory.COMPETITION,
    }
    return aliases.get(normalized, TargetCategory.OWNED)


def normalize_targets(targets: list[Target]) -> list[Target]:
    normalized: list[Target] = []
    seen: set[str] = set()
    for target in targets:
        value = str(target.value or "").strip()
        if not value:
            continue
        key = value.casefold()
        if key in seen:
            continue
        seen.add(key)
        normalized.append(
            Target(
                value=value,
                match_mode=target.match_mode,
                label=str(target.label or "").strip(),
                category=parse_target_category(target.category),
                brand_regex=str(target.brand_regex or "").strip(),
            )
        )
    return normalized


def validate_targets(targets: list[Target]) -> list[Target]:
    normalized = normalize_targets(targets)
    if not normalized:
        raise ValueError("Add at least one property to monitor.")
    if len(normalized) > MAX_MONITOR_PROPERTIES:
        raise ValueError(f"At most {MAX_MONITOR_PROPERTIES} properties are supported.")
    for target in normalized:
        compile_brand_regex(target.brand_regex or None)
    return normalized


def _property_retrieved(
    run: GroundingRun,
    target: Target,
    *,
    retrieval_complete: bool,
) -> ObservationState:
    if any(target.value in source.target_matches for source in run.sources):
        return ObservationState.YES
    return ObservationState.NO if retrieval_complete else ObservationState.UNKNOWN


def _property_cited(
    run: GroundingRun,
    target: Target,
    *,
    citation_complete: bool,
    resolve_redirects: bool,
    has_failed_redirects: bool,
) -> ObservationState:
    if any(target.value in citation.target_matches for citation in run.citations):
        return ObservationState.YES
    if (
        target.match_mode is MatchMode.URL_PREFIX
        and resolve_redirects
        and has_failed_redirects
    ):
        return ObservationState.UNKNOWN
    return ObservationState.NO if citation_complete else ObservationState.UNKNOWN


def compute_property_results(
    run: GroundingRun,
    request: GroundingRequest,
    *,
    retrieval_complete: bool = False,
    citation_complete: bool = True,
) -> list[PropertyResult]:
    resolve_redirects = should_resolve_citation_redirects(request)
    has_failed_redirects = any(
        is_grounding_redirect_url(citation.url)
        and citation.metadata.get("redirect_resolution") == "failed"
        for citation in run.citations
    )
    results: list[PropertyResult] = []
    for target in request.targets:
        brand_state, brand_matches = match_brand(run.response_text, target.brand_regex or None)
        results.append(
            PropertyResult(
                value=target.value,
                label=display_label(target),
                category=target.category,
                retrieved=_property_retrieved(
                    run,
                    target,
                    retrieval_complete=retrieval_complete,
                ),
                cited=_property_cited(
                    run,
                    target,
                    citation_complete=citation_complete,
                    resolve_redirects=resolve_redirects,
                    has_failed_redirects=has_failed_redirects,
                ),
                brand_mentioned=brand_state,
                brand_matches=tuple(brand_matches),
            )
        )
    return results


def aggregate_target_state(
    results: list[PropertyResult],
    field: str,
    *,
    prefer_category: TargetCategory | None = None,
) -> ObservationState:
    if not results:
        return ObservationState.UNKNOWN
    scoped = (
        [item for item in results if item.category is prefer_category]
        if prefer_category is not None
        else results
    ) or results
    states = [getattr(item, field) for item in scoped]
    if any(state is ObservationState.YES for state in states):
        return ObservationState.YES
    if any(state is ObservationState.UNKNOWN for state in states):
        return ObservationState.UNKNOWN
    if all(state is ObservationState.NOT_APPLICABLE for state in states):
        return ObservationState.NOT_APPLICABLE
    return ObservationState.NO


def property_results_to_dict(results: list[PropertyResult]) -> list[dict[str, object]]:
    return [
        {
            "value": item.value,
            "label": item.label,
            "category": item.category.value,
            "retrieved": item.retrieved.value,
            "cited": item.cited.value,
            "brandMentioned": item.brand_mentioned.value,
            "brandMatches": list(item.brand_matches),
        }
        for item in results
    ]
